README.md
# FauxmoESP

Amazon Alexa support for ESP8266 and ESP32 devices.

This is a library for ESP8266/ESP32-based devices that emulates Philips Hue lights and thus allows you to control them using this protocol, in particular from Alexa-powered devices like the Amazon Echo or the Dot.

[![version](https://img.shields.io/badge/version-3.2-brightgreen.svg)](CHANGELOG.md)
[![codacy](https://img.shields.io/codacy/grade/44478ddd58fe4cc6a2bc5598232663b8/master.svg)](https://www.codacy.com/app/xoseperez/fauxmoesp/dashboard)
[![license](https://img.shields.io/badge/license-MIT-orange.svg)](LICENSE)

[![donate](https://img.shields.io/badge/donate-PayPal-blue.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=pjvint%40gmail%2ecom&lc=US&no_note=0&currency_code=USD&bn=PP%2dDonationsBF%3abtn_donate_LG%2egif%3aNonHostedGuest)

## NEW! Discussions on Github
We now have a place for general discussion, ideas, projects etc. - https://github.com/vintlabs/fauxmoESP/discussions/133

## Notice

**As of October 2020 https://github.com/vintlabs/fauxmoESP is the new home for fauxmoESP!!**

**Many thanks for all of the work that Xose Perez has put into this project!**

*I have migrated all of the issues from the old repo on Bitbucket here to GitHub, and I have closed many stale issues. If I have closed an issue that you feel should still be open, feel free to reopen it or submit a new one.*


## History

2020-12-22 Version 3.2 released. Devices now show properly in the Alexa App as a bulb.

2020-12-08 Version 3.1.2 released. New version available in Arduino Library Manager or at https://github.com/vintlabs/fauxmoESP/releases/tag/3.1.2  PlatformIO is still pending update.

Before version 3.0.0, the library used a different protocol (emulating Belkin Wemo devices). The library was a port of Maker Musings' [Fauxmo Python library][6] to the ESP8266 platform. Support for ESP32 and Gen2 devices was added by Frank Hellmann <frank at vfx dot to> and Bibi Blocksberg respectively.

Since version 3.0.0 the library uses a different approach and emulates Philips Hue lights instead. This allows for a simpler code and also support for numeric values (you can now say "Alexa, set light to 50"). This version has been inspired by the [node-red-contrib-alexa-local](https://github.com/originallyus/node-red-contrib-alexa-local) plugin for NodeRED and the [ESPalexa](https://github.com/Aircoookie/Espalexa) library by Christian Schwinne.

## Dependencies

Besides the libraries already included with the Arduino Core for ESP8266 or ESP32, these libraries are also required to use fauxmoESP:

ESP8266:

* This library uses [ESPAsyncTCP][3] library by [me-no-dev][5]

ESP32:

* This library uses [AsyncTCP][4] library by [me-no-dev][5]

### PlatformIO

If you are using PlatformIO (check the section bellow on how to compile it) the required libraries should be installed automatically.

### Arduino IDE

You will need to install the required library from sources. Your best option is to download the library as a ZIP file and install it using the option under "Sketch > Include Library > Add .ZIP Library...".

You can look for it manually but I have gathered the URL here for convenience:

|Device|Library|Repository|ZIP|
|-|-|-|-|
|ESP8266|**ESPAsyncTCP** by Hristo Gochkov ESP8266|[GIT](https://github.com/me-no-dev/ESPAsyncTCP)|[ZIP](https://github.com/me-no-dev/ESPAsyncTCP/archive/master.zip)|
|ESP32|**AsyncTCP** by Hristo Gochkov ESP32|[GIT](https://github.com/me-no-dev/AsyncTCP)|[ZIP](https://github.com/me-no-dev/AsyncTCP/archive/master.zip)|

## Usage

The library is very easy to use, basically instantiate an object, connect to the Wifi, add one or more virtual devices and bind the callback to get the messages. An schematic example could be:

```C
#include <fauxmoESP.h>

fauxmoESP fauxmo;

void setup() {

    Serial.begin(115200);

    ... connect to wifi ...

    fauxmo.addDevice("light one");
    fauxmo.addDevice("light two");
    fauxmo.addDevice("light three");
    fauxmo.addDevice("light four");

    fauxmo.setPort(80); // required for gen3 devices
    fauxmo.enable(true);

    fauxmo.onSetState([](unsigned char device_id, const char * device_name, bool state, unsigned char value) {
        Serial.printf("[MAIN] Device #%d (%s) state: %s value: %d\n", device_id, device_name, state ? "ON" : "OFF", value);
    });

}

void loop() {
    fauxmo.handle();
}

```

(Check the examples folder)

## To use with ESP-IDF

Add `#include "Arduino.h"`


Then run the "discover devices" option from your Alexa app or web (in the Smart Home section). A new device with the name you have configured should appear. Tell Alexa to switch it on or off and check your terminal ;)

## Troubleshooting

Current status of the library:

|Platform|Gen1|Gen2|Gen3|
|---|---|---|---|
|ESP8266 Core 2.3.X|OK|OK?|OK (1)|
|ESP8266 Core 2.4.0|OK (2)|OK? (2)|OK (1, 2)|
|ESP8266 Core 2.4.1|OK (2)|OK? (2)|OK (1, 2)|
|ESP8266 Core 2.4.2|OK (2)|OK? (2)|OK (1, 2)|
|ESP32|OK|OK?|OK (1)|

(1) When using gen3 devices TCP port must be 80 always.
(2) Arduino Core for ESP8266 requires LwIP set to "v1.4 Higher Bandwidth".

* fauxmoESP 3.1.X: When using with gen3 devices TCP port must be 80. You can define it with the `setPort` method.

* fauxmoESP 3.1.X: If you application already uses port 80 you can prevent fauxmoESP from creating its own webserver and inject the values from your application handlers to the library. Check the fauxmoESP_External_Server example.

* fauxmoESP 3.X.X: When using Arduino Core for ESP8266 v2.4.X, double check you are building the project with LwIP Variant set to "v1.4 Higher Bandwidth". You can change it from the Tools menu in the Arduino IDE or passing the `-DPIO_FRAMEWORK_ARDUINO_LWIP_HIGHER_BANDWIDTH` build flag to PlatformIO.

* fauxmoESP 2.X.X: The `onGetState` method accepts a function (a callback) that will be called when a new message arrives. Try not to do many things inside the callback, it should return as fast as possible. Instead of adding logic here just save the data (device_id and state, for instance) and process it from your main loop.

* Some people have reported problems when the ESP and the Alexa devices are connected to different wireless networks (like 2.4 and 5GHz bands on some routers). See https://bitbucket.org/xoseperez/fauxmoesp/issues/53.

* Latest version of ESP Async Webserver fails building with the current setup. Use version 1.2.2 max. See fauxmoESP_External_Server example `platformio.ini` file.

[1]:https://github.com/esp8266/Arduino
[2]:http://docs.platformio.org/en/stable/platforms/espressif8266.html#using-arduino-framework-with-staging-version
[3]:https://github.com/me-no-dev/ESPAsyncTCP
[4]:https://github.com/me-no-dev/AsyncTCP
[5]:https://github.com/me-no-dev
[6]:https://github.com/makermusings/fauxmo

## If you enjoy this, please consider supporting us by purchasing a module from us!
http://www.vintlabs.com

## License

Copyright (C) 2016-2020 by Xose Pérez <xose dot perez at gmail dot com>, 2020 by Paul Vint <pjvint at gmail dot com>

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR

CHANGELOG.md
# fauxmoESP change log

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [3.2] 2020-12-22
### Changed
Fixed modelid so devices properly show as a light in the Alexa App

## [3.1.2] 2020-12-08
### Changed
- Changed `uniqueid` to fix Alexa discovery issue 
- Added `setDeviceUniqueId(unsigned char id, const char *uniqueid)` to optionally manually set a uniqueid

## [3.1.1] 2020-03-21
### Changed
- Freeze platformio.ini configuration

### Added 
- Add ssdp:discover to accepted UDP answers (#58, thanks to @rodymary)

## [3.1.0] 2018-11-18
### Fixed
- PlatformIO dependencies (#60)
- Fix issue with enable() call that prevented it from working the first time (#39)

### Changed
- Improved examples & documentation (#62, #65)
- Not using printf_P in ESP8266 since 2.3.0 does not support it (#69)

### Added
- Support for gen3 (third generation) Alexa devices (#66)
- Option to use external webserver to handle TCP requests
- RX timeout
- Free class resources upon object deletion
- Method to remove a device
- Method to rename a devices based on the old name (#57)
- Method to get a device_id from the name (#57)
- Method setState to report state changes to Alexa (#59)

## [3.0.2] 2018-10-08
### Fixed
- Removed symlinks to allow Arduino track the library

### Changed
- Debug messages moved to PROGMEME

## [3.0.1] 2018-08-29
### Fixed
- Possible segmentation fault in getDeviceName

## [3.0.0] 2018-08-15
### Changed
- Completely new paradigm, emulating Philips Hue API instead of Belkin Wemo API
- onSetState signature has changed to accommodate the brightness value
- onGetState no longer exists

### Added
- DEBUG_FAUXMO_VERBOSE setting for verbose debug output (defaults to false)

## [2.4.3] 2018-06-23
Dummy version to force PlatformIO parsing

## [2.4.2] 2018-01-14
### Fixed
- Removed a Serial.print

## [2.4.1] 2018-01-10
### Fixed
- Fixed dependency issue in PlatformIO (the hard way)

## [2.4.0] 2018-01-05
### Fixed
- Fixed examples for Arduino IDE (#34)

### Added
- Support for ESP32 (#28, thanks to Frank Hellmann)
- Preliminary support for Gen2 devices (thanks to Bibi Blocksberg)

## [2.3.0] 2017-11-08
### Fixed
- Answer correctly to GetBinaryState queries (thanks to John McC)

### Deprecated
- Use onSetState callback instead of onMessage callback
- Use onGetState callback instead of setState method

## [2.2.1] 2017-10-25
### Fixed
- Only change state if request is a SetBinaryState action

## [2.2.0] 2017-09-01
### Fixed
- Remove dependency on ESPAsyncWebServer

### Added
- Option to rename devices

## [2.1.1] 2017-08-25
### Fixed
- Call UDP beginMulticast on onStationModeGotIP

## [2.1.0] 2017-02-15
### Added
- ESPAsyncWebServer to manage TCP connections

### Fixed
- Memory leaks in UDP and TCP connections

## [2.0.0] 2017-01-05
### Added
- Different discovery strategies to maximize device discovery
- UDP polling required (call handle() in your loop)

### Changed
- ```onMessage``` callback signature changed to add device_id

### Removed
- Removed dependency on ESPAsyncUDP for compatibility with current stable release of Arduino Code for ESP8266

## [1.0.0] 2016-11-26
### Added
- Support for multiple virtual devices

### Changed
- ```onMessage``` callback signature has changed to allow passing the device name

### Deprecated
- ```setDeviceName``` no longer exists, use ```addDevice``` instead

## [0.1.0] 2016-11-21
- Initial working version


src/fauxmoESP.cpp
/*

FAUXMO ESP

Copyright (C) 2016-2020 by Xose Pérez <xose dot perez at gmail dot com>

The MIT License (MIT)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/

#include <Arduino.h>
#include "fauxmoESP.h"

// -----------------------------------------------------------------------------
// UDP
// -----------------------------------------------------------------------------

void fauxmoESP::_sendUDPResponse() {

	DEBUG_MSG_FAUXMO("[FAUXMO] Responding to M-SEARCH request\n");

	IPAddress ip = WiFi.localIP();
    String mac = WiFi.macAddress();
    mac.replace(":", "");
    mac.toLowerCase();

	char response[strlen(FAUXMO_UDP_RESPONSE_TEMPLATE) + 128];
    snprintf_P(
        response, sizeof(response),
        FAUXMO_UDP_RESPONSE_TEMPLATE,
        ip[0], ip[1], ip[2], ip[3],
		_tcp_port,
        mac.c_str(), mac.c_str()
    );

	#if DEBUG_FAUXMO_VERBOSE_UDP
    	DEBUG_MSG_FAUXMO("[FAUXMO] UDP response sent to %s:%d\n%s", _udp.remoteIP().toString().c_str(), _udp.remotePort(), response);
	#endif

    _udp.beginPacket(_udp.remoteIP(), _udp.remotePort());
	#if defined(ESP32)
	    _udp.printf(response);
	#else
	    _udp.write(response);
	#endif
    _udp.endPacket();

}

void fauxmoESP::_handleUDP() {

	int len = _udp.parsePacket();
    if (len > 0) {

		unsigned char data[len+1];
        _udp.read(data, len);
        data[len] = 0;

		#if DEBUG_FAUXMO_VERBOSE_UDP
			DEBUG_MSG_FAUXMO("[FAUXMO] UDP packet received\n%s", (const char *) data);
		#endif

        String request = (const char *) data;
        if (request.indexOf("M-SEARCH") >= 0) {
            if ((request.indexOf("ssdp:discover") > 0) || (request.indexOf("upnp:rootdevice") > 0) || (request.indexOf("device:basic:1") > 0)) {
                _sendUDPResponse();
            }
        }
    }

}


// -----------------------------------------------------------------------------
// TCP
// -----------------------------------------------------------------------------

void fauxmoESP::_sendTCPResponse(AsyncClient *client, const char * code, char * body, const char * mime) {

	char headers[strlen_P(FAUXMO_TCP_HEADERS) + 32];
	snprintf_P(
		headers, sizeof(headers),
		FAUXMO_TCP_HEADERS,
		code, mime, strlen(body)
	);

	#if DEBUG_FAUXMO_VERBOSE_TCP
		DEBUG_MSG_FAUXMO("[FAUXMO] Response:\n%s%s\n", headers, body);
	#endif

	client->write(headers);
	client->write(body);

}

String fauxmoESP::_deviceJson(unsigned char id) {

	if (id >= _devices.size()) return "{}";

	fauxmoesp_device_t device = _devices[id];

  DEBUG_MSG_FAUXMO("[FAUXMO] Sending device info for \"%s\", uniqueID = \"%s\"\n", device.name, device.uniqueid);
  char buffer[strlen_P(FAUXMO_DEVICE_JSON_TEMPLATE) + 64];
  snprintf_P(
        buffer, sizeof(buffer),
        FAUXMO_DEVICE_JSON_TEMPLATE,
        device.name, device.uniqueid,
        device.state ? "true": "false",
        device.value, device.colormode,
        device.hue, device.saturation, 
        device.ct
    );

	return String(buffer);

}

String fauxmoESP::_byte2hex(uint8_t zahl)
{
  String hstring = String(zahl, HEX);
  if (zahl < 16)
  {
    hstring = "0" + hstring;
  }

  return hstring;
}

String fauxmoESP::_makeMD5(String text)
{
  unsigned char bbuf[16];
  String hash = "";
  MD5Builder md5;
  md5.begin();
  md5.add(text);
  md5.calculate();
  
  md5.getBytes(bbuf);
  for (uint8_t i = 0; i < 16; i++)
  {
    hash += _byte2hex(bbuf[i]);
  }

  return hash;
}

bool fauxmoESP::_onTCPDescription(AsyncClient *client, String url, String body) {

	(void) url;
	(void) body;

	DEBUG_MSG_FAUXMO("[FAUXMO] Handling /description.xml request\n");

	IPAddress ip = WiFi.localIP();
    String mac = WiFi.macAddress();
    mac.replace(":", "");
    mac.toLowerCase();

	char response[strlen_P(FAUXMO_DESCRIPTION_TEMPLATE) + 64];
    snprintf_P(
        response, sizeof(response),
        FAUXMO_DESCRIPTION_TEMPLATE,
        ip[0], ip[1], ip[2], ip[3], _tcp_port,
        ip[0], ip[1], ip[2], ip[3], _tcp_port,
        mac.c_str(), mac.c_str()
    );

	_sendTCPResponse(client, "200 OK", response, "text/xml");

	return true;

}

bool fauxmoESP::_onTCPList(AsyncClient *client, String url, String body) {

	DEBUG_MSG_FAUXMO("[FAUXMO] Handling list request\n");

	// Get the index
	int pos = url.indexOf("lights");
	if (-1 == pos) return false;

	// Get the id
	unsigned char id = url.substring(pos+7).toInt();

	// This will hold the response string	
	String response;

	// Client is requesting all devices
	if (0 == id) {

		response += "{";
		for (unsigned char i=0; i< _devices.size(); i++) {
			if (i>0) response += ",";
			response += "\"" + String(i+1) + "\":" + _deviceJson(i);
		}
		response += "}";

	// Client is requesting a single device
	} else {
		response = _deviceJson(id-1);
	}

	_sendTCPResponse(client, "200 OK", (char *) response.c_str(), "application/json");
	
	return true;

}

bool fauxmoESP::_onTCPControl(AsyncClient *client, String url, String body) {

	// "devicetype" request
	if (body.indexOf("devicetype") > 0) {
		DEBUG_MSG_FAUXMO("[FAUXMO] Handling devicetype request\n");
		_sendTCPResponse(client, "200 OK", (char *) "[{\"success\":{\"username\": \"2WLEDHardQrI3WHYTHoMcXHgEspsM8ZZRpSKtBQr\"}}]", "application/json");
		return true;
	}

	// "state" request
	if ((url.indexOf("state") > 0) && (body.length() > 0)) {

		// Get the index
		int pos = url.indexOf("lights");
		if (-1 == pos) return false;

		DEBUG_MSG_FAUXMO("[FAUXMO] Handling state request\n");

		// Get the index
		unsigned char id = url.substring(pos+7).toInt();
		if (id > 0) {

			--id;

			// Brightness
			pos = body.indexOf("bri");
			if (pos > 0) {
				unsigned char value = body.substring(pos+5).toInt();
				_devices[id].value = value;
				_devices[id].state = (value > 0);
        _setRGBFromHSV(id);
			} else if (body.indexOf("false") > 0) {
      	_devices[id].state = false;
			} else {
				_devices[id].state = true;
				if (0 == _devices[id].value) 
				{
				  _devices[id].value = 254;
          _setRGBFromHSV(id);
				}
			}

      // Hue / Saturation
      pos = body.indexOf("hue");
      if (pos > 0)
      {
        unsigned int hue = body.substring(pos + 5).toInt();
        DEBUG_MSG_FAUXMO("[FAUXMO] Setting hue to %d\n", hue);
        _devices[id].hue = hue;
        strcpy(_devices[id].colormode, "hs");
      }
      
      pos = body.indexOf("\"sat\"");
      if (pos > 0)
      {
        unsigned char saturation = body.substring(pos + 6).toInt();
        DEBUG_MSG_FAUXMO("[FAUXMO] Setting saturation to %d\n", saturation);
        _devices[id].saturation = saturation;
        strcpy(_devices[id].colormode, "hs");
        _setRGBFromHSV(id);
      }
printf("Checking CT\n");
      // Colour temperature
      pos = body.indexOf("\"ct\"");
      if (pos > 0)
      {
        //printf("Found CT\n");
        unsigned int ct = body.substring(pos + 5).toInt();
        DEBUG_MSG_FAUXMO("[FAUXMO] Setting ct to %d\n", ct);
        _devices[id].ct = ct;
        strcpy(_devices[id].colormode, "ct");
        _setRGBFromCT(id);
      }
      
			char response[strlen_P(FAUXMO_TCP_STATE_RESPONSE)+23];
			snprintf_P(
				response, sizeof(response),
				FAUXMO_TCP_STATE_RESPONSE,
				id+1, _devices[id].state ? "true" : "false", id+1, _devices[id].value, id+1, _devices[id].hue, id+1, _devices[id].saturation, id+1,  _devices[id].ct
			);
      //DEBUG_MSG_FAUXMO("[FAUXMO] Sending state response:\n%s\n", response);
			_sendTCPResponse(client, "200 OK", response, "text/xml");

			if (_setCallback) {
				_setCallback(id, _devices[id].name, _devices[id].state, _devices[id].value, _devices[id].hue, _devices[id].saturation, _devices[id].ct);
			}

			return true;

		}

	}

	return false;
	
}

bool fauxmoESP::_onTCPRequest(AsyncClient *client, bool isGet, String url, String body) {

    if (!_enabled) return false;

	#if DEBUG_FAUXMO_VERBOSE_TCP
		DEBUG_MSG_FAUXMO("[FAUXMO] isGet: %s\n", isGet ? "true" : "false");
		DEBUG_MSG_FAUXMO("[FAUXMO] URL: %s\n", url.c_str());
		if (!isGet) DEBUG_MSG_FAUXMO("[FAUXMO] Body:\n%s\n", body.c_str());
	#endif

	if (url.equals("/description.xml")) {
        return _onTCPDescription(client, url, body);
    }

	if (url.startsWith("/api")) {
		if (isGet) {
			return _onTCPList(client, url, body);
		} else {
       		return _onTCPControl(client, url, body);
		}
	}

	return false;

}

bool fauxmoESP::_onTCPData(AsyncClient *client, void *data, size_t len) {

    if (!_enabled) return false;

	char * p = (char *) data;
	p[len] = 0;

	#if DEBUG_FAUXMO_VERBOSE_TCP
		DEBUG_MSG_FAUXMO("[FAUXMO] TCP request\n%s\n", p);
	#endif

	// Method is the first word of the request
	char * method = p;

	while (*p != ' ') p++;
	*p = 0;
	p++;
	
	// Split word and flag start of url
	char * url = p;

	// Find next space
	while (*p != ' ') p++;
	*p = 0;
	p++;

	// Find double line feed
	unsigned char c = 0;
	while ((*p != 0) && (c < 2)) {
		if (*p != '\r') {
			c = (*p == '\n') ? c + 1 : 0;
		}
		p++;
	}
	char * body = p;

	bool isGet = (strncmp(method, "GET", 3) == 0);

	return _onTCPRequest(client, isGet, url, body);

}

void fauxmoESP::_onTCPClient(AsyncClient *client) {

	if (_enabled) {

	    for (unsigned char i = 0; i < FAUXMO_TCP_MAX_CLIENTS; i++) {

	        if (!_tcpClients[i] || !_tcpClients[i]->connected()) {

	            _tcpClients[i] = client;

	            client->onAck([i](void *s, AsyncClient *c, size_t len, uint32_t time) {
	            }, 0);

	            client->onData([this, i](void *s, AsyncClient *c, void *data, size_t len) {
	                _onTCPData(c, data, len);
	            }, 0);
	            client->onDisconnect([this, i](void *s, AsyncClient *c) {
			if(_tcpClients[i] != NULL) {
	                    _tcpClients[i]->free();
	                    _tcpClients[i] = NULL;
	                }
			else {
	                    DEBUG_MSG_FAUXMO("[FAUXMO] Client %d already disconnected\n", i);
	                }
	                delete c;
	                DEBUG_MSG_FAUXMO("[FAUXMO] Client #%d disconnected\n", i);
	            }, 0);

	            client->onError([i](void *s, AsyncClient *c, int8_t error) {
	                DEBUG_MSG_FAUXMO("[FAUXMO] Error %s (%d) on client #%d\n", c->errorToString(error), error, i);
	            }, 0);

	            client->onTimeout([i](void *s, AsyncClient *c, uint32_t time) {
	                DEBUG_MSG_FAUXMO("[FAUXMO] Timeout on client #%d at %i\n", i, time);
	                c->close();
	            }, 0);

                    client->setRxTimeout(FAUXMO_RX_TIMEOUT);

	            DEBUG_MSG_FAUXMO("[FAUXMO] Client #%d connected\n", i);
	            return;

	        }

	    }

		DEBUG_MSG_FAUXMO("[FAUXMO] Rejecting - Too many connections\n");

	} else {
		DEBUG_MSG_FAUXMO("[FAUXMO] Rejecting - Disabled\n");
	}

    client->onDisconnect([](void *s, AsyncClient *c) {
        c->free();
        delete c;
    });
    client->close(true);

}

void fauxmoESP::_setRGBFromHSV(unsigned char id) 
{
  if (id < 0) 
    return;

    float dh, ds, dv;
    dh = _devices[id].hue;
    ds = _devices[id].saturation;
    dv = _devices[id].value / 256.0;

    // lifted from https://github.com/Aircoookie/Espalexa/blob/master/src/EspalexaDevice.cpp    
    float h = ((float)dh)/65536.0;
    float s = ((float)ds)/255.0;
    byte i = floor(h*6);
    float f = h * 6-i;
    float p = 255 * (1-s);
    float q = 255 * (1-f*s);
    float t = 255 * (1-(1-f)*s);
    switch (i%6) {
      case 0: 
        _devices[id].red = 255;
        _devices[id].green = t;
        _devices[id].blue = p;
        break;
      case 1: 
        _devices[id].red = q;
        _devices[id].green = 255;
        _devices[id].blue = p;
        break;
      case 2: 
        _devices[id].red = p;
        _devices[id].green = 255;
        _devices[id].blue = t;
        break;
      case 3: 
        _devices[id].red = p;
        _devices[id].green = q;
        _devices[id].blue = 255;
        break;
      case 4: 
        _devices[id].red = t;
        _devices[id].green = p;
        _devices[id].blue = 255;
        break;
      case 5: 
        _devices[id].red = 255;
        _devices[id].green = p;
        _devices[id].blue = q;
        break;
    }

    _devices[id].red = _devices[id].red * dv;
    _devices[id].green = _devices[id].green * dv;
    _devices[id].blue = _devices[id].blue * dv;
    
 }

void fauxmoESP::_setRGBFromCT(unsigned char id) 
{
  if (id < 0) 
    return;
  float temp = 10000.0 / _devices[id].ct;
  float r, g, b;

  if (temp <= 66)
  {
    r = 255;
    g = 99.470802 * log(temp) - 161.119568;

    if (temp <= 19)
    {
      b = 0;
    }
    else
    {
      b = 138.517731 * log(temp - 10) - 305.044793;
    }
  }
  else
  {
    r = 329.698727 * pow(temp - 60, -0.13320476);
    g = 288.12217 * pow(temp - 60, -0.07551485 );
    b = 255;
  }

  r = constrain(r, 0, 255);
  g = constrain(g, 0, 255);
  b = constrain(b, 0, 255);

  _devices[id].red = r;
  _devices[id].green = g;
  _devices[id].blue = b;

  //printf("RGB %f %f %f\n", r, g, b);    
   
}

// -----------------------------------------------------------------------------
// Devices
// -----------------------------------------------------------------------------

fauxmoESP::~fauxmoESP() {
  	
	// Free the name for each device
	for (auto& device : _devices) {
		free(device.name);
  	}
  	
	// Delete devices  
	_devices.clear();

}

void fauxmoESP::setDeviceUniqueId(unsigned char id, const char *uniqueid)
{
    strncpy(_devices[id].uniqueid, uniqueid, FAUXMO_DEVICE_UNIQUE_ID_LENGTH);
}

unsigned char fauxmoESP::addDevice(const char * device_name) {

    fauxmoesp_device_t device;
    unsigned int device_id = _devices.size();

    // init properties
    device.name = strdup(device_name);
  	device.state = false;
	  device.value = 0;
    device.hue = 0;
    device.saturation = 0;
    device.ct = 0;
    strcpy(device.colormode, "hs");

    // create the uniqueid
    String mac = WiFi.macAddress();
    mac.replace(":", "");
    mac.toLowerCase();

    // To create the uniqueID, we use the device's MAC, add the device name
    // then use a truncated MD5 hash to create an ID that is unique and repeatable
    mac.concat(device.name); 
    String hash = _makeMD5(mac).substring(0,FAUXMO_DEVICE_UNIQUE_ID_LENGTH);  
    strcpy(device.uniqueid, hash.c_str());
    
    // Attach
    _devices.push_back(device);

    DEBUG_MSG_FAUXMO("[FAUXMO] Device '%s' added as #%d\n", device_name, device_id);

    return device_id;

}

int fauxmoESP::getDeviceId(const char * device_name) {
    for (unsigned int id=0; id < _devices.size(); id++) {
        if (strcmp(_devices[id].name, device_name) == 0) {
            return id;
        }
    }
    return -1;
}

bool fauxmoESP::renameDevice(unsigned char id, const char * device_name) {
    if (id < _devices.size()) {
        free(_devices[id].name);
        _devices[id].name = strdup(device_name);
        DEBUG_MSG_FAUXMO("[FAUXMO] Device #%d renamed to '%s'\n", id, device_name);
        return true;
    }
    return false;
}

bool fauxmoESP::renameDevice(const char * old_device_name, const char * new_device_name) {
	int id = getDeviceId(old_device_name);
	if (id < 0) return false;
	return renameDevice(id, new_device_name);
}

bool fauxmoESP::removeDevice(unsigned char id) {
    if (id < _devices.size()) {
        free(_devices[id].name);
		_devices.erase(_devices.begin()+id);
        DEBUG_MSG_FAUXMO("[FAUXMO] Device #%d removed\n", id);
        return true;
    }
    return false;
}

bool fauxmoESP::removeDevice(const char * device_name) {
	int id = getDeviceId(device_name);
	if (id < 0) return false;
	return removeDevice(id);
}

char * fauxmoESP::getDeviceName(unsigned char id, char * device_name, size_t len) {
    if ((id < _devices.size()) && (device_name != NULL)) {
        strncpy(device_name, _devices[id].name, len);
    }
    return device_name;
}

char * fauxmoESP::getColormode(unsigned char id, char * cm, size_t len)
{
   if (id < _devices.size())
   {
      strncpy(cm, _devices[id].colormode, len);
   }
   
   return cm;
}

uint8_t fauxmoESP::getRed(unsigned char id)
{
  return _devices[id].red;
}
uint8_t fauxmoESP::getGreen(unsigned char id)
{
  return _devices[id].green;
}
uint8_t fauxmoESP::getBlue(unsigned char id)
{
  return _devices[id].blue;
}

// For on/off and Brightness
bool fauxmoESP::setState(unsigned char id, bool state, unsigned char value) {
    if (id < _devices.size()) {
		_devices[id].state = state;
		_devices[id].value = value;
		return true;
	}
	return false;
}

bool fauxmoESP::setState(const char * device_name, bool state, unsigned char value) {
	int id = getDeviceId(device_name);
	if (id < 0) return false;
	_devices[id].state = state;
	_devices[id].value = value;
	return true;
}

// For hue / Saturation
bool fauxmoESP::setState(unsigned char id, bool state, unsigned int hue, unsigned int saturation) {
    if (id < _devices.size()) {
      _devices[id].hue = hue;
      _devices[id].saturation = saturation;
    return true;
  }
  return false;
}

bool fauxmoESP::setState(const char * device_name, bool state, unsigned int hue, unsigned int saturation) {
  int id = getDeviceId(device_name);
  if (id < 0) return false;
    _devices[id].hue = hue;
    _devices[id].saturation = saturation;
  return true;
}

// For Colour Temperature (ct)
bool fauxmoESP::setState(unsigned char id, bool state, unsigned int ct) {
    if (id < _devices.size()) {
      _devices[id].ct = ct;
      _setRGBFromCT(id);
      return true;
  }
  return false;
}

bool fauxmoESP::setState(const char * device_name, bool state, unsigned int ct) {
  int id = getDeviceId(device_name);
  if (id < 0) return false;
  {
    _devices[id].ct = ct;
    _setRGBFromCT(id);
    return true;
  }
  
}



// -----------------------------------------------------------------------------
// Public API
// -----------------------------------------------------------------------------

bool fauxmoESP::process(AsyncClient *client, bool isGet, String url, String body) {
	return _onTCPRequest(client, isGet, url, body);
}

void fauxmoESP::handle() {
    if (_enabled) _handleUDP();
}

void fauxmoESP::enable(bool enable) {

	if (enable == _enabled) return;
    _enabled = enable;
	if (_enabled) {
		DEBUG_MSG_FAUXMO("[FAUXMO] Enabled\n");
	} else {
		DEBUG_MSG_FAUXMO("[FAUXMO] Disabled\n");
	}

    if (_enabled) {

		// Start TCP server if internal
		if (_internal) {
			if (NULL == _server) {
				_server = new AsyncServer(_tcp_port);
				_server->onClient([this](void *s, AsyncClient* c) {
					_onTCPClient(c);
				}, 0);
			}
			_server->begin();
		}

		// UDP setup
		#ifdef ESP32
            _udp.beginMulticast(FAUXMO_UDP_MULTICAST_IP, FAUXMO_UDP_MULTICAST_PORT);
        #else
            _udp.beginMulticast(WiFi.localIP(), FAUXMO_UDP_MULTICAST_IP, FAUXMO_UDP_MULTICAST_PORT);
        #endif
        DEBUG_MSG_FAUXMO("[FAUXMO] UDP server started\n");

	}

}
